code/obj/item/pet_carrier.dm

```diff
@@ -0,0 +1,319 @@
+ * This was created without access to the pre-existing chicken carrier code so it could be pretty bad.
+ *
+ * The given mobs that a carrier is allowed to hold is determined by var/mob/allowed_mobs which should be overridden by any child types of the
+ * carrier. The maximum capacity is determined by var/carrier_max_capacity which can also be overridden.
+ *
+ * Whether a mob can break out or not is determined by var/can_break_out, and if FALSE, any occupant cannot escape the carrier on their own.
+ *
+ * There is support for spawning with an animal already inside by overriding var/mob/default_mob with a valid path.
+ *
+ * The icon for the carrier is constructed using a base where an overlay of the front of the carrier and two /obj/dummys are layered on top.
+ * These dummy objects are held in the carrier's vis_contents and each of them hold the mobs and the door of the carrier. The reason why the
+ * vis_contents of vis_contents_proxy hold the mobs rather than the carrier doing so directly is so that an alpha mask can prevent parts of the
+ * occupants from rendering outside of the inside of the carrier.
+ *
+ * For any mobs inside, the atmosphere they'll attempt to breathe from is taken from the current turf that the carrier occupies, even if the carrier
+ * is being held in someone's hand.
+ */
+/obj/item/pet_carrier
+	name = "pet carrier"
+	desc = "A surprisingly roomy carrier for transporting small animals."
+	icon = 'icons/obj/items/pet_carrier.dmi'
+	icon_state = "carrier-full"
+	item_state = "carrier-open"
+	w_class = W_CLASS_BULKY
+
+	/// Please override this in child types to specify what can actually fit in.
+	var/mob/allowed_mobs = /mob/living/critter/small_animal
+	/// If FALSE, an occupant cannot escape the carrier on their own.
+	var/can_break_out = TRUE
+	/// How many mobs can fit inside the crate. Usually not overridden by anything, this is to let the system be permissive for var-editing.
+	var/carrier_max_capacity = DEFAULT_MAX_CAPACITY
+	/// If not null, the pet carrier will spawn with one of this mob on New().
+	var/mob/default_mob = null
+
+	/// The icon_state for the src.TRAP_MOB() actionbar.
+	var/trap_mob_icon_state = "carrier-full"
+	/// The icon_state for the src.RELEASE_MOB() actionbar.
+	var/release_mob_icon_state = "carrier-full-open"
+	// Alpha mask icon state for cutting out the mob on non-transparent pixels.
+	var/carrier_alpha_mask = "carrier-mask"
+
+	// Empty carrier icon state name.
+	var/empty_carrier_icon_state = "carrier"
+
+	// Grate icon state names.
+	var/grate_open_icon_state = "grate-open"
+	var/grate_closed_icon_state = "grate-closed"
+
+	// Carrier item state names.
+	var/carrier_open_item_state = "carrier-open"
+	var/carrier_closed_item_state = "carrier-closed"
+
+	/// Carrier-related (grate_proxy, vis_contents_proxy) vis_flags.
+	var/carrier_vis_flags = VIS_INHERIT_ID | VIS_INHERIT_LAYER | VIS_INHERIT_PLANE
+	/// Mob-specific vis_flags.
+	var/mob_vis_flags = VIS_INHERIT_LAYER | VIS_INHERIT_PLANE
```

Comment:
Lot of vars here that aren't ever changed or are only referenced in one spot, I would think they could just be removed, or made constants